File: backend/vizql/format_resolver.py
# backend/vizql/format_resolver.py
"""Plan 10a — memoised layered format resolver.

Walk order (most-specific first):
    Mark > Field > Worksheet > Data Source > Workbook > default.

Memoisation keyed on (mark_id, field_id, sheet_id, ds_id, prop).
"""
from __future__ import annotations

import logging
from functools import lru_cache
from typing import Iterable, Mapping, Optional

from vizql.formatting_types import (
    DataSourceSelector,
    FieldSelector,
    MarkSelector,
    Selector,
    StyleProp,
    StyleRule,
    WorkbookSelector,
    WorksheetSelector,
)

logger = logging.getLogger(__name__)

# Allowed primitive types for StyleRule property values. Anything else
# (callables, custom objects, None sentinels) is rejected at ingest to keep
# memoisation safe and serialisation deterministic.
_ALLOWED_VALUE_TYPES = (str, int, float, bool, tuple)


class ResolverError(ValueError):
    """Raised for invalid rule shapes or invalid resolve arguments."""


class FormatResolver:
    def __init__(
        self,
        rules: Iterable[StyleRule],
        *,
        cache_enabled: bool = True,
        cache_maxsize: int = 4096,
    ) -> None:
        self._cache_enabled = cache_enabled
        self._cache_maxsize = cache_maxsize
        self.update_rules(rules)

    # --- Public API ------------------------------------------------------

    def update_rules(self, rules: Iterable[StyleRule]) -> None:
        flat = list(rules)
        self._validate(flat)
        # Pre-bucket rules by selector kind → (key → rule) for O(1) lookup.
        self._by_mark: dict[str, list[StyleRule]] = {}
        self._by_field: dict[str, list[StyleRule]] = {}
        self._by_sheet: dict[str, list[StyleRule]] = {}
        self._by_ds: dict[str, list[StyleRule]] = {}
        self._workbook: list[StyleRule] = []
        for rule in flat:
            self._bucket(rule)
        self._rules = flat
        self._reset_cache()

    def resolve(
        self,
        mark_id: Optional[str],
        field_id: Optional[str],
        sheet_id: Optional[str],
        ds_id: Optional[str],
        prop: StyleProp,
        *,
        default: object = None,
    ) -> object:
        if not isinstance(prop, StyleProp):
            raise ResolverError(f"resolve: prop must be StyleProp, got {type(prop).__name__}")
        if self._cache_enabled:
            return self._resolve_cached(mark_id, field_id, sheet_id, ds_id, prop, default)
        return self._resolve_uncached(mark_id, field_id, sheet_id, ds_id, prop, default)

    def resolve_all(
        self,
        mark_id: Optional[str],
        field_id: Optional[str],
        sheet_id: Optional[str],
        ds_id: Optional[str],
    ) -> Mapping[StyleProp, object]:
        out: dict[StyleProp, object] = {}
        for prop in StyleProp:
            val = self.resolve(mark_id, field_id, sheet_id, ds_id, prop, default=_UNSET)
            if val is not _UNSET:
                out[prop] = val
        return out

    def resolve_all_with_source(
        self,
        mark_id: Optional[str],
        field_id: Optional[str],
        sheet_id: Optional[str],
        ds_id: Optional[str],
    ) -> Mapping[StyleProp, tuple[object, str]]:
        out: dict[StyleProp, tuple[object, str]] = {}
        for prop in StyleProp:
            result = self._resolve_uncached_with_source(
                mark_id, field_id, sheet_id, ds_id, prop
            )
            if result is not None:
                out[prop] = result
        return out

    def cache_info(self) -> dict[str, int]:
        if not self._cache_enabled:
            return {"hits": 0, "misses": 0, "maxsize": 0, "currsize": 0}
        info = self._resolve_cached.cache_info()
        return {
            "hits": info.hits,
            "misses": info.misses,
            "maxsize": info.maxsize,
            "currsize": info.currsize,
        }

    # --- Internal --------------------------------------------------------

    def _validate(self, rules: list[StyleRule]) -> None:
        for rule in rules:
            if not isinstance(rule, StyleRule):
                raise ResolverError(f"expected StyleRule, got {type(rule).__name__}")
            for prop, value in rule.properties.items():
                if not isinstance(prop, StyleProp):
                    raise ResolverError(f"property key must be StyleProp, got {prop!r}")
                if not isinstance(value, _ALLOWED_VALUE_TYPES):
                    raise ResolverError(
                        f"property value must be primitive (str/int/float/bool/tuple); got {type(value).__name__}"
                    )

    def _bucket(self, rule: StyleRule) -> None:
        s = rule.selector
        if isinstance(s, MarkSelector):
            self._by_mark.setdefault(s.mark_id, []).append(rule)
        elif isinstance(s, FieldSelector):
            self._by_field.setdefault(s.field_id, []).append(rule)
        elif isinstance(s, WorksheetSelector):
            self._by_sheet.setdefault(s.sheet_id, []).append(rule)
        elif isinstance(s, DataSourceSelector):
            self._by_ds.setdefault(s.ds_id, []).append(rule)
        elif isinstance(s, WorkbookSelector):
            self._workbook.append(rule)
        else:
            raise ResolverError(f"unknown selector type: {type(s).__name__}")

    def _reset_cache(self) -> None:
        if self._cache_enabled:
            @lru_cache(maxsize=self._cache_maxsize)
            def _inner(mark_id, field_id, sheet_id, ds_id, prop, default):
                return self._resolve_uncached(mark_id, field_id, sheet_id, ds_id, prop, default)
            self._resolve_cached = _inner

    def _layer_chain(
        self, mark_id, field_id, sheet_id, ds_id
    ) -> list[tuple[str, list[StyleRule]]]:
        # Walk most-specific → least-specific.
        return [
            ("mark", self._by_mark.get(mark_id, []) if mark_id else []),
            ("field", self._by_field.get(field_id, []) if field_id else []),
            ("sheet", self._by_sheet.get(sheet_id, []) if sheet_id else []),
            ("ds", self._by_ds.get(ds_id, []) if ds_id else []),
            ("workbook", self._workbook),
        ]

    def _resolve_uncached(self, mark_id, field_id, sheet_id, ds_id, prop, default):
        for _layer, bucket in self._layer_chain(mark_id, field_id, sheet_id, ds_id):
            # Multiple rules at same layer → last wins (most recent override).
            for rule in reversed(bucket):
                if prop in rule.properties:
                    return rule.properties[prop]
        return default

    def _resolve_uncached_with_source(self, mark_id, field_id, sheet_id, ds_id, prop):
        for layer, bucket in self._layer_chain(mark_id, field_id, sheet_id, ds_id):
            for rule in reversed(bucket):
                if prop in rule.properties:
                    return rule.properties[prop], layer
        return None


_UNSET = object()

File: frontend/src/components/dashboard/freeform/lib/formatResolver.ts
// Plan 10a — TS port of backend/vizql/format_resolver.py.
// CONTRACT: bit-for-bit identical output to Python resolver for the
// fixtures under backend/vizql/tests/fixtures/format_parity.
import {
  BOOL_STYLE_PROPS,
  NUMERIC_STYLE_PROPS,
  type Selector,
  StyleProp,
  type StyleRule,
  type StyleValue,
} from './formattingTypes';

export type ResolveResult<T = StyleValue> = { value: T; layer: string } | null;

const LAYER_ORDER: ReadonlyArray<'mark' | 'field' | 'sheet' | 'ds' | 'workbook'> = [
  'mark',
  'field',
  'sheet',
  'ds',
  'workbook',
];

export class FormatResolverError extends Error {}

export class FormatResolver {
  private byMark = new Map<string, StyleRule[]>();
  private byField = new Map<string, StyleRule[]>();
  private bySheet = new Map<string, StyleRule[]>();
  private byDs = new Map<string, StyleRule[]>();
  private workbook: StyleRule[] = [];
  private cache = new Map<string, StyleValue | undefined>();
  private hits = 0;
  private misses = 0;

  constructor(rules: readonly StyleRule[], private readonly cacheEnabled = true) {
    this.updateRules(rules);
  }

  updateRules(rules: readonly StyleRule[]) {
    this.byMark.clear(); this.byField.clear();
    this.bySheet.clear(); this.byDs.clear();
    this.workbook = [];
    for (const rule of rules) this.bucket(rule);
    this.cache.clear();
    this.hits = 0;
    this.misses = 0;
  }

  resolve(
    markId: string | null,
    fieldId: string | null,
    sheetId: string | null,
    dsId: string | null,
    prop: StyleProp,
    defaultValue?: StyleValue,
  ): StyleValue | undefined {
    const key = `${markId}|${fieldId}|${sheetId}|${dsId}|${prop}`;
    if (this.cacheEnabled && this.cache.has(key)) {
      this.hits += 1;
      const cached = this.cache.get(key);
      return cached === undefined ? defaultValue : cached;
    }
    this.misses += 1;
    const found = this.walk(markId, fieldId, sheetId, dsId, prop);
    const value = found?.value;
    if (this.cacheEnabled) this.cache.set(key, value);
    return value === undefined ? defaultValue : value;
  }

  resolveWithSource(
    markId: string | null, fieldId: string | null,
    sheetId: string | null, dsId: string | null,
    prop: StyleProp,
  ): ResolveResult | null {
    return this.walk(markId, fieldId, sheetId, dsId, prop);
  }

  resolveAll(
    markId: string | null, fieldId: string | null,
    sheetId: string | null, dsId: string | null,
  ): Partial<Record<StyleProp, StyleValue>> {
    const out: Partial<Record<StyleProp, StyleValue>> = {};
    for (const prop of Object.values(StyleProp)) {
      const v = this.resolve(markId, fieldId, sheetId, dsId, prop as StyleProp);
      if (v !== undefined) out[prop as StyleProp] = v;
    }
    return out;
  }

  cacheInfo() {
    return { hits: this.hits, misses: this.misses, size: this.cache.size };
  }

  // --- internal -------------------------------------------------------

  private bucket(rule: StyleRule) {
    const s = rule.selector as Selector;
    switch (s.kind) {
      case 'mark': this.push(this.byMark, s.markId, rule); break;
      case 'field': this.push(this.byField, s.fieldId, rule); break;
      case 'sheet': this.push(this.bySheet, s.sheetId, rule); break;
      case 'ds': this.push(this.byDs, s.dsId, rule); break;
      case 'workbook': this.workbook.push(rule); break;
      default: throw new FormatResolverError(`unknown selector: ${JSON.stringify(s)}`);
    }
  }
  private push(m: Map<string, StyleRule[]>, k: string, r: StyleRule) {
    const arr = m.get(k); if (arr) arr.push(r); else m.set(k, [r]);
  }

  private walk(
    markId: string | null, fieldId: string | null,
    sheetId: string | null, dsId: string | null, prop: StyleProp,
  ): ResolveResult | null {
    const bucketFor = (layer: string): StyleRule[] => {
      if (layer === 'mark') return (markId && this.byMark.get(markId)) || [];
      if (layer === 'field') return (fieldId && this.byField.get(fieldId)) || [];
      if (layer === 'sheet') return (sheetId && this.bySheet.get(sheetId)) || [];
      if (layer === 'ds') return (dsId && this.byDs.get(dsId)) || [];
      return this.workbook;
    };
    for (const layer of LAYER_ORDER) {
      const bucket = bucketFor(layer);
      for (let i = bucket.length - 1; i >= 0; i -= 1) {
        const rule = bucket[i];
        if (prop in rule.properties) {
          const raw = rule.properties[prop];
          return { value: coerce(prop, raw), layer };
        }
      }
    }
    return null;
  }
}

export function coerce(prop: StyleProp, raw: StyleValue): StyleValue {
  if (BOOL_STYLE_PROPS.has(prop)) return typeof raw === 'boolean' ? raw : raw === 'true' || raw === true;
  if (NUMERIC_STYLE_PROPS.has(prop)) return typeof raw === 'number' ? raw : Number(raw);
  return raw;
}
